crates/wiggle/borrow/src/lib.rs

```diff
@@ -84,17 +99,30 @@ impl InnerBorrowChecker {
         Ok(h)
     }
 
-    fn borrow(&mut self, r: Region) -> Result<BorrowHandle, GuestError> {
-        if self.is_borrowed(r) {
+    fn shared_borrow(&mut self, r: Region) -> Result<BorrowHandle, GuestError> {
+        if self.is_mut_borrowed(r) {
             return Err(GuestError::PtrBorrowed(r));
         }
         let h = self.new_handle()?;
-        self.borrows.insert(h, r);
+        self.shared_borrows.insert(h, r);
         Ok(h)
     }
 
-    fn unborrow(&mut self, h: BorrowHandle) {
-        let _ = self.borrows.remove(&h);
+    fn mut_borrow(&mut self, r: Region) -> Result<BorrowHandle, GuestError> {
+        if self.is_shared_borrowed(r) || self.is_mut_borrowed(r) {
+            return Err(GuestError::PtrBorrowed(r));
+        }
+        let h = self.new_handle()?;
+        self.mut_borrows.insert(h, r);
+        Ok(h)
+    }
+
+    fn shared_unborrow(&mut self, h: BorrowHandle) {
+        let _ = self.shared_borrows.remove(&h);
```

Comment:
Is it possible to `debug_assert!` that this did a removal?

Reply:
Great idea